crypto/cms/cms_env.c

```diff
@@ -779,111 +875,207 @@ int CMS_RecipientInfo_encrypt(CMS_ContentInfo *cms, CMS_RecipientInfo *ri)
 
 /* Check structures and fixup version numbers (if necessary) */
 
-static void cms_env_set_originfo_version(CMS_EnvelopedData *env)
+static int cms_env_get_oi_version(CMS_OriginatorInfo *org, int version)
 {
-    CMS_OriginatorInfo *org = env->originatorInfo;
     int i;
+
     if (org == NULL)
-        return;
+        return version;
+
     for (i = 0; i < sk_CMS_CertificateChoices_num(org->certificates); i++) {
         CMS_CertificateChoices *cch;
         cch = sk_CMS_CertificateChoices_value(org->certificates, i);
         if (cch->type == CMS_CERTCHOICE_OTHER) {
-            env->version = 4;
-            return;
+            return 4;
         } else if (cch->type == CMS_CERTCHOICE_V2ACERT) {
-            if (env->version < 3)
-                env->version = 3;
+            if (version < 3)
+                return 3;
         }
     }
 
     for (i = 0; i < sk_CMS_RevocationInfoChoice_num(org->crls); i++) {
         CMS_RevocationInfoChoice *rch;
         rch = sk_CMS_RevocationInfoChoice_value(org->crls, i);
         if (rch->type == CMS_REVCHOICE_OTHER) {
-            env->version = 4;
-            return;
+            return 4;
         }
     }
+
+    return version;
 }
 
-static void cms_env_set_version(CMS_EnvelopedData *env)
+static int cms_env_get_ris_version(STACK_OF(CMS_RecipientInfo) *ris,
+                                   int version)
 {
     int i;
     CMS_RecipientInfo *ri;
 
+    if (version >= 3)
+        return version;
+
+    for (i = 0; i < sk_CMS_RecipientInfo_num(ris); i++) {
+        ri = sk_CMS_RecipientInfo_value(ris, i);
+        if (ri->type == CMS_RECIPINFO_PASS || ri->type == CMS_RECIPINFO_OTHER) {
+            return 3;
+        } else if (ri->type != CMS_RECIPINFO_TRANS
+                   || ri->d.ktri->version != 0) {
+            return 2;
+        }
+    }
+
+    return version;
+}
+
+static void cms_env_set_version(CMS_EnvelopedData *env)
+{
     /*
      * Can't set version higher than 4 so if 4 or more already nothing to do.
      */
     if (env->version >= 4)
         return;
 
-    cms_env_set_originfo_version(env);
+    env->version = cms_env_get_oi_version(env->originatorInfo, env->version);
+    env->version = cms_env_get_ris_version(env->recipientInfos, env->version);
```

Comment:
Yeah I'm sorry about this. I remember that I thought that there are different rules but after I refactored it and read again the RFC, I found out that it's actually the same so I just kept the refactored version as I thought that it's a bit more readable. Should I revert this refactoring?

Reply:
Unless there is a good reason for changing this code, I'd prefer to keep the churn to a minimum. So, reverting this seems like the way to go.

Reply:
I have been looking into this again and when reading it again and figuring out the logic, the new version seems much clearer to me - I was able to get the logic of the new version generation quicker. So the main reason to keep it would be a better clarity of the code. Would you still prefer to put it back as it was? Potentially I could create a new PR for this later if that's better?

Reply:
After thinking a new PR with a proper test would be better so put it back as it was.